base/broadcast.jl

```diff
@@ -261,7 +261,16 @@ Base.@propagate_inbounds function Base.iterate(bc::Broadcasted, s)
     return (bc[i], (s[1], newstate))
 end
 
-Base.IteratorSize(::Type{<:Broadcasted{<:Any,<:NTuple{N,Base.OneTo}}}) where {N} = Base.HasShape{N}()
+Base.IteratorSize(::Type{T}) where {T<:Broadcasted} = Base.HasShape{ndims(T)}()
+Base.ndims(BC::Type{<:Broadcasted{<:Any,Nothing}}) = _maxndims(fieldtype(BC, 2))
+function Base.ndims(BC::Type{<:Broadcasted{<:AbstractArrayStyle{N},Nothing}}) where {N}
+    N isa Integer && return N
+    _maxndims(fieldtype(BC, 2))
+end
+Base.@pure _maxndims(T) = mapfoldl(_ndims, max, fieldtypes(T))
```

Comment:
See `Iterators.zip_iteratorsize` for and example of how this is normally implemented

Reply:
looks nested zip also have problem?
```julia
julia> a = Iterators.zip(1:10,1:10)
zip(1:10, 1:10)

julia> b = zip(a, a);

julia> c = zip(b, b);

julia> @code_warntype Base.IteratorSize(c)
MethodInstance for Base.IteratorSize(::Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}, Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}}})
  from Base.IteratorSize(x) in Base at generator.jl:92
Arguments
  #self#::Core.Const(Base.IteratorSize)
  x::Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}, Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}}}
Body::Any
1 ─ %1 = Base.typeof(x)::Core.Const(Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}, Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}}})
│   %2 = Base.IteratorSize(%1)::Any
└──      return %2


julia> @code_warntype Base.IteratorSize(b)
MethodInstance for Base.IteratorSize(::Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}})
  from Base.IteratorSize(x) in Base at generator.jl:92
Arguments
  #self#::Core.Const(Base.IteratorSize)
  x::Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}}
Body::Base.HasShape{1}
1 ─ %1 = Base.typeof(x)::Core.Const(Base.Iterators.Zip{Tuple{Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}, Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}}})
│   %2 = Base.IteratorSize(%1)::Core.Const(Base.HasShape{1}())
└──      return %2


julia> @code_warntype Base.IteratorSize(a)
MethodInstance for Base.IteratorSize(::Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}})
  from Base.IteratorSize(x) in Base at generator.jl:92
Arguments
  #self#::Core.Const(Base.IteratorSize)
  x::Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}}
Body::Base.HasShape{1}
1 ─ %1 = Base.typeof(x)::Core.Const(Base.Iterators.Zip{Tuple{UnitRange{Int64}, UnitRange{Int64}}})
│   %2 = Base.IteratorSize(%1)::Core.Const(Base.HasShape{1}())
└──      return %2
```

Reply:
I've tried implementing this as in `Iterators.zip_iteratorsize` but can't get nested broadcasts to pass [the `@inferred` test](https://github.com/JuliaLang/julia/pull/44061/files#diff-5e385f8c1fb4092e5aca743c367485dc9e0cd1e8e2558ea16d7b1012efcc4523R885).  Is there a reason `@pure` shouldn't be used here @vtjnash ?  I'm not familiar with when it's safe or not.

Reply:
`@pure` should not be used. It is never safe.

Reply:
I've removed `@pure` and instead defined methods of `_maxndims` for small tuples which has helped the inference on nested broadcasts.  

The inference through nested broadcasts won't work for more complex cases e.g. where the original `broadcasted` and the nested broadcast have >2 args.  For example a test like this would fail:
```
bc = Base.broadcasted(+, AD1(randn(3)), AD1(randn(3)), AD1(randn(3)))
bc_nest = Base.broadcasted(+, bc , bc, bc)
@test @inferred(Base.IteratorSize(bc_nest)) === Base.HasShape{1}()
```
My thinking was that perhaps we could live with more complex cases like this being uninferrable, so long as the simpler cases can be inferred.